syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.routeguide";
option java_outer_classname = "RouteGuideProto";
option objc_class_prefix = "RTG";

package shopping;


service Market {
  rpc CheckConnection(PingRequest) returns (Response) {}
  rpc RegisterSeller(Seller) returns (Response) {}
  rpc SellItem(Item) returns (RegisterItemResponse) {}
  rpc UpdateItem(UpdateItemRequest) returns (Response) {}
  rpc DeleteItem(DeleteItemRequest) returns (Response) {}
  rpc showItems(ShowItemRequest) returns (stream Item) {}
  
  rpc RegisterBuyer(Buyer) returns (Response) {}
  rpc SearchItem(DisplayItemRequest) returns (stream ItemForDisplay) {}
  rpc buyItem(BuyItemRequest) returns (Response) {}
  rpc addToWishlist(WishListRequest) returns (Response) {}
  rpc rateItem(RateItemRequest) returns (Response) {}
}

service Buyernotifyication {
  rpc notifyyConnection(PingRequest) returns (Response) {}
  rpc notifyyItemUpdated(ItemUpdatednotifyication) returns (Response) {}
}

service Sellernotifyication {
  rpc notifyyConnection(PingRequest) returns (Response) {}
  rpc notifyyItemBought(ItemBoughtnotifyication) returns (Response) {}
}

message Item {
  string seller_uuid = 1;
  string item_id = 2;
  string name = 3;
  string category= 4;
  int32 price = 5;
  int32 quantity= 6;
  float rating = 7;
  string description = 8;
}

message ItemForDisplay {
  string item_id = 1;
  string name = 2;
  string category= 3;
  int32 price = 4;
  float rating = 5;
  string description = 6;
}

message Response {
  bool status = 1;
  string message = 2;
}

message RegisterItemResponse {
  bool status = 1;
  string message = 2;
  string Item_id = 3;
}


message DeleteItemRequest {
  string seller_uuid = 1;
  string Item_id = 2;
}

message ShowItemRequest {
  string seller_uuid = 1;
}

message DisplayItemRequest {
  string buyer_uuid = 1;
  string Item_name = 2;
  string Item_category = 3;
}

message Buyer {
  string uuid = 1;
  string notify_server_address = 2;
}

message Seller {
  string uuid = 1;
  string notify_server_address = 2;
}

message BuyItemRequest {
  string buyer_uuid = 1;
  string Item_id = 2;
  int32 quantity= 3;
}

message WishListRequest {
  string buyer_uuid = 1;
  string Item_id = 2;
}


message UpdateItemRequest {
  string seller_uuid = 1;
  string Item_id = 2;
  Item new_Item = 3;
}

message RateItemRequest {
  string buyer_uuid = 1;
  string Item_id = 2;
  float rating = 3;
}



message ItemBoughtnotifyication {
  string Item_id = 1;
  int32 quantity= 2;
}

message ItemUpdatednotifyication {
  ItemForDisplay new_Item = 1;
}

message PingRequest {
  string message = 1;
}


